src/Course/State.hs

```diff
@@ -94,17 +77,13 @@ instance Functor (State s) where
 -- >>> runState (State (\s -> ((+3), s P.++ ["apple"])) <*> State (\s -> (7, s P.++ ["banana"]))) []
 -- (10,["apple","banana"])
 instance Applicative (State s) where
-  pure ::
-    a
-    -> State s a
-  pure =
-    error "todo: Course.State pure#instance (State s)"
-  (<*>) ::
-    State s (a -> b)
-    -> State s a
-    -> State s b 
-  (<*>) =
-    error "todo: Course.State (<*>)#instance (State s)"
+  pure :: a -> State s a
+  pure x = State (\y -> (x, y))
+  (<*>) :: State s (a -> b) -> State s a -> State s b
+  (<*>) s1 s2 = State (\x -> (e1 x, e2 x))
+    where
+      e1 x = eval s1 x (eval s2 x)
+      e2 x = exec s2 (exec s1 x)
```

Comment:
So... the `s1` state has the function inside it... and the body of `e1` is evaluating s1, then applying that to `x` and the evaluation of s2 with x.

It's a bit tricky to read.... here's an alternative which you might find easier to understand:

```
    State (\s -> -- still needs to be inside a function on some data.. "s" here
      let
        (f, s2) = runState stateF s -- run the state to get the f out of stateF with the data value "s"
        (x, s3) = runState stateX s2 -- thread that last resultant data value "s2" back in to runState to get the "x" value out, that we'll apply "f" to in a second...
      in
        (f x, s3) -- the tuple we reply with has the "f" extracted from the first state applied to the "x" extracted from the second state, and the resultant value is the threaded state that we got back from running state to get "x" out (ie the cumulative effected result value).
    )
```

WDYT?

Reply:
Also worth noting, we don't have to do the `(f, s2)` first, we could choose the `x` to do with `s2` first if we wanted.

Reply:
I definitely like the `let` and the destructuring instead of `fst`, `eval`, or `exec`.